README.md
# contactlessElevator v1

How to set up RaspberryPi

    Programming Raspbian Os on SD card

        Insert the microSD card in the card reader
        Setting the SD card in your Raspberry Pi: https://projects.raspberrypi.org/en/projects/raspberry-pi-setting-up/2
    
    
    Setting the Raspberry Pi for use
        
        Using GUI, monitor, keyboard and mouse
            Insert the microSD card in the card slot
            Connect the micro HDMI the the monitor slot near the power (left) and the other HDMI end to a monitor
            Connect the USB cable of wire and mouse (or bluetooth dongle for wireless) to the USB slots shown as Keyboard and Mouse
            Connect type C power cable (5V 3A or 5V 1A rating) to the power connector
            Turn the power on and the RaspberryPi is ready to be used
            For internet access, connect the Pi to Wifi or insert a LAN cable in Network slot
            

Details of all the prototype scripts

    a directory labeled model which contains the AI models for speech to text algorithm which the VOSK library uses. 
        Do Not rename or delete this folder, will render the project useless.
    controlLib.py which is a collection of classes and functions used to generate control signals to the first and second floor buttons
    micStreaming.py which is a collection of classes and functions to capture the audio stream, convert it to text and use the controlLib functions to generate the control signals.
    wrapperScript.sh which is a shell script to keep the python script running in case of some fault or expedition killing the python process.
        If the RaspberryPi shutdown, the shell script will also shutdown.
    If you want the wrapperScript to start auto-running on powerup, explore this option. Google has many tutorials on how to do this.
    


Installing dependencies and codes for prototype
    
    Raspbian OS is a linux based operating system and supports most of the Linux commands on it. Start the terminal on RaspberryPi, type one at a time and press enter to run. Below are the steps:
        sudo apt-get install python3-pip python3-dev
        git clone https://github.com/alphacep/vosk-api.git
        cd vosk-api/python
        python3 setup.py
        sudo pip3 install vosk
        pip3 install PyAudio
        cd ~ (Go back to home directory)
        gh repo clone shivam18shah/contactlessElevator (will get all the code files)
        (Other dependencies if missing will be printed on terminal while running the code and should be installed accordingly)
        

Setting up crontab jobs for script automation / Starting scripts without crontab
    Execute the following command in a terminal
        crontab -l : to check for any scheduled cron jobs
    Confirm that nothing has been scheduled yet
        crontab - e : then choose a text editor once it asks for it 
    In the editor tab type 5 * * * * source ~/finalCodes/wrapperScript.sh
    Note:
        there is a space after last asterisk and ./
        check any of the above links to check what is the format for the 5 asterisk which is min, hr, day of the month, month, day of the week
    Save and close the editor which will ensure that the wrapperScript runs every 5 min.
    
    Further research can be done to improve it, you can start the script by going to the finalCodes folder (type cd ~/finalCodes) and then type source wrapperScript.sh command in the terminal. Thus, following this step should start the code.
    Since the script will be running in the background, if you want to kill the script, then
        type ps aux in terminal
        look for the PID (it will be number) of micStreaming in the list printed
        type kill PID (replace PID with actual PID number of script) to stop the script
        To restart the script, again type the command above in bold.
    Please note that the links for reference materials are included at the end of this README
    

Final installation
    The final installation is to be done using the RaspberryPi installation kit screws. The following instructions should be taken:
        Holes can be drilled on the acrylic board in similar arrangement as above and then the components and be held in place using the installation kit screws.
        Remember to keep the pins of the component (4) in the diagram above facing upwards else, you won't be able to make the pin connections.
        Male to female or female to female connectors (included in the components ordered) can be used to make all the other  connections (except the one colored red in the circuit diagram) wherever required.
        Anything specific to the pinout diagram of individual diagrams eg (solid state relay - to understand which pin is which (total 4 pins on ssd)) can be found by Google search of the specific part.

    Common practices and suggestions:
        Check individual parts before making the final connection, check voltages using multimeter to check if the voltages and currents are in range and do not damage the components.
        You can type the command “pinout” (without double quotes) in the RaspberryPi terminal to get the pin diagram of RaspberryPi 4 GPIOs.
        Remember to turn off the Wifi connection of RaspberryPi after programming to avoid it getting tampered with. It is a requirement that RPi remains offline.
        Put a login and password to login to RaspberryPi (videos can be found on Google) to block random access to RaspberryPi.
        Do not pull the power USB from RPi unless you shutdown the RaspberryPi from GUI or using terminal command.
        
        
Starting the prototype after installation
    After the RaspberryPi has been connected to the elevator panel, it's time to start the crontab job to start the prototype.
    When you are downloading the codes on RaspberryPi, don't forget to turn on the SSH connection from the RaspberryPi config (Google how to).
    To do this,you can connect an ethernet cable from RaspberryPi to your laptop ethernet port.
    Open a command line application like MobaXterm or Putty and open its terminal.
    Type ssh raspberrypi.local (it's important that SSH has to be turned on on RPi)
    Type the password in and you will get access to the command line terminal of RPi where you can follow Step 4 in the tutorial and get the crontab job running
    This should ensure the prototype starts running.
    Test the prototype before and after closing the elevator panel.


Giving the voice commands:
    Trigger commands: Hey Elevator, Okay Elevator, Hi Elevator
    Floor Commands: First Floor, Second Floor


Useful references:
    Linux cron jobs: https://www.youtube.com/watch?v=QZJ1drMQz1A
    Cron jobs every 5 minutes: https://crontab.guru/every-5-minutes
    Periodic cron jobs: https://www.thegeekstuff.com/2011/07/cron-every-5-minutes/
    Using cron jobs to run scripts automatically: https://www.jessicayung.com/automate-running-a-script-using-crontab/
    Scheduling a python script or cron job: https://gavinwiener.medium.com/how-to-schedule-a-python-script-cron-job-dea6cbf69f4e
    Raspberry Pi documentation: https://www.raspberrypi.org/documentation/remote-access/ssh/
    Raspberry Pi setup: https://learn.adafruit.com/raspberry-pi-computer-quick-start
    


testCodes/history.txt
    1  ifconfig
    2  sudo raspi-config
    3  ks
    4  ls
    5  clear
    6  python3
    7  clear
    8  ls
    9  mkdir EPICS
   10  cd EPICS/
   11  mkdir testCodes
   12  cd testCodes/
   13  ls
   14  c ~
   15  cd ~
   16  ls
   17  sudo apt-get intall spyder
   18  sudo apt-get install spyder
   19  cd EPICS/testCodes/
   20  ls
   21  spyder &
   22  cd EPICS/testCodes/
   23  vim LEDControl.py
   24  sudo apt-get install vi
   25  sudo apt-get install vim
   26  cd ~
   27  sudo apt-get install vim
   28  sudo apt install vim
   29  cd EPICS/testCodes/
   30  python ledControl.py 
   31  pinout
   32  spyder &
   33  cd EPICS/testCodes/
   34  ls
   35  history | grep python
   36  python ledControl.py 
   37  vim ledControl.py 
   38  ls
   39  clear
   40  spyder &
   41  python ledControl.py 
   42  python switchLED.py 
   43  clear
   44  python switchLED.py 
   45  clear
   46  ls
   47  python ledControl.py 
   48  python switchLED.py 
   49  clear
   50  python ledControl.py 
   51  python switchLED.py 
   52  clear
   53  python switchLED.py 
   54  python ledControl.py 
   55  python switchLED.py 
   56  python ledControl.py 
   57  python switchLED.py 
   58  python ledControl.py 
   59  python switchLED.py 
   60  clear
   61  python ledControl.py 
   62  clear
   63  python switchLED.py 
   64  clear
   65  python ledControl.py 
   66  python switchLED.py 
   67  python ledControl.py 
   68  clear
   69  python switchLED.py 
   70  python ledControl.py 
   71  clear
   72  python switchLED.py 
   73  python ledControl.py 
   74  clear
   75  cd EPICS/testCodes/
   76  ls
   77  spyder &
   78  python ledControl.py 
   79  clear
   80  python fourledControl.py 
   81  clear
   82  python fourledControl.py 
   83  clear
   84  python fourledControl.py 
   85  clear
   86  python fourledControl.py 
   87  clear
   88  clar
   89  clear
   90  spyder &
   91  clear
   92  cd EPICS/testCodes/
   93  python ledControl.py 
   94  python commandLED.py 
   95  python commandLED.py 
   96  python ledControl.py 
   97  clear
   98  python commandLED.py 
   99  python ledControl.py 
  100  clear
  101  python commandLED.py 
  102  python ledControl.py 
  103  clear
  104  python commandLED.py 
  105  python ledControl.py 
  106  clear
  107  pinout
  108  clear
  109  sudo apt-get install i2c-tools
  110  sudo apt--get install python-smbus
  111  sudo apt-get install python-smbus
  112  i2cdetect -y 1
  113  clear
  114  spyder &
  115  clear
  116  i2cdetect -y 1
  117  clear
  118  cd EPICS/testCodes/
  119  python lcdTest.py 
  120  i2cdetect -y 1
  121  clear
  122  python lcdTest.py 
  123  sudo raspi-config
  124  clear
  125  i2cdetect -y 1
  126  clear
  127  python lcdTest.py 
  128  clear
  129  python lcdTest.py 
  130  clear
  131  python lcdTest.py 
  132  clear
  133  history | grep -i "i2c"
  134  i2cdetect -y 1 sudo apt-get install i2c-tools
  135  history | grep -i "apt-gt"
  136  history | grep -i "apt-get"
  137  i2cdetect -y 0
  138  i2cdetect -y 1
  139  clear
  140  i2cdetect -y 1
  141  i2cdetect -y 0
  142  i2cdetect -y 2
  143  clear
  144  i2cdetect -y 1
  145  clear
  146  python lcdTest.py 
  147  clear
  148  spyder &
  149  git --version
  150  clear
  151  git config --global user.name "SarthakGadre"
  152  git config --global user.email "sugadre@asu.edu"
  153  git config --list
  154  clear
  155  git config --help
  156  clear
  157  ls
  158  cd EPICS
  159  ls
  160  cd testCodes/
  161  ls
  162  cd ..
  163  clear
  164  ls
  165  ls -la
  166  clear
  167  git init
  168  ls
  169  ls -al
  170  clear
  171  ls
  172  git status
  173  ls
  174  vim gitignore
  175  cat gitignore 
  176  mv gitignore ./.gitignore
  177  ls
  178  ls .al
  179  ls -al
  180  vim .gitignore 
  181  clear
  182  ls
  183  git status
  184  git add ./testCodes/ledControl.py 
  185  git status
  186  git add ./testCodes/switchLED.py 
  187  git add ./testCodes/lcdTest.py 
  188  git add ./testCodes/fourledControl.py 
  189  git add ./testCodes/commandLED.py 
  190  ls
  191  git add ./testCodes/I2C_LCD_driver.py 
  192  ls
  193  cd testCodes/
  194  ls
  195  cd ..
  196  git add .gitignore 
  197  git status 
  198  ls
  199  clear
  200  git status 
  201  git reset testCodes/I2C_LCD_driver.py
  202  git reset
  203  git status
  204  git add .gitignore 
  205  git add testCodes/ledControl.py 
  206  git status
  207  git commit -m "First Code to control LED using GPIO"
  208  git sattus
  209  git status
  210  clear
  211  git log
  212  clear
  213  git branch
  214  git branch -a
  215  clear
  216  ls
  217  git pull
  218  cd ~
  219  clear
  220  git config
  221  git config --list
  222  clear
  223  git status
  224  cd EPICS/
  225  ls
  226  git status
  227  git pull
  228  git branch
  229  git push
  230  git remote add origin htts://github.com/SarthakGadre/contactlessElevator.git
  231  git status
  232  git branch
  233  git branch -M ain
  234  git branch -M main
  235  git pull
  236  git remote add origin https://github.com/SarthakGadre/contactlessElevator.git
  237  git remote remove origin
  238  git remote add origin https://github.com/SarthakGadre/contactlessElevator.git
  239  git pull
  240  git branch -M main
  241  git push -u origin main
  242  clear
  243  ls
  244  clear
  245  cd EPICS/
  246  cd testCodes/
  247  ls
  248  spyder &
  249  vim testIP.txt
  250  cat testIP.txt 
  251  clear
  252  gedit testIP.txt &
  253  vim testIP.txt &
  254  vim testIP.txt
  255  clear
  256  cd EPICS/testCodes/
  257  ls
  258  clear
  259  spyder &
  260  cd EPICS/
  261  ls
  262  git status
  263  git config
  264  git config --global
  265  git config --local
  266  clear
  267  git config --local
  268  clear
  269  git status
  270  git add testCodes/testIP.txt testCodes/floorControl.py testCodes/commandControl.py testCodes/commandTest.txt 
  271  git status
  272  git commit 

testCodes/lcdTest.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 13 20:07:29 2020

@author: pi
"""

# =============================================================================
# Header file for interfacing i2c controlled 16x2 display
# Need to be included for controlling the display
# Connect SDC to GPIO 2 and SCL to GPIO 3 which is I2C 1 iterface of RPi
# =============================================================================
import I2C_LCD_driver
import time


# =============================================================================
# Assigning all the functions under class lcd to the pointer mylcd
# =============================================================================
mylcd = I2C_LCD_driver.lcd()


# =============================================================================
# Printing a simple string on the display
#("String to be printed",row number(1,2), column number (0-15))
# =============================================================================
#mylcd.lcd_display_string("a",1,0)
#time.sleep(5)


# =============================================================================
# Command to clear the display
# =============================================================================

mylcd.lcd_clear()

# =============================================================================
# Printing the date and time format
# =============================================================================
#
mylcd.lcd_display_string("Time: %s" %time.strftime("%H:%M:%S"), 1)
mylcd.lcd_display_string("Date: %s" %time.strftime("%m/%d/%Y"), 2)


time.sleep(3)
mylcd.lcd_clear()

# =============================================================================
# Macro to scroll the text from left to right once
# =============================================================================
str_pad = " " * 16
my_long_string = "This is a long string that needs to be dislayed!"
my_long_string = str_pad + my_long_string

for i in range (0, len(my_long_string)):
    lcd_text = my_long_string[i:(i+16)]
    mylcd.lcd_display_string(lcd_text,1)
    time.sleep(0.4)
    mylcd.lcd_display_string(str_pad,1)

# =============================================================================
# Macro to scroll text from right to left once
# =============================================================================
#padding = " " * 16
#my_long_string = "To be or not to be that is the question!"
#padded_string = padding+ my_long_string + padding
#
#for i in range (0, len(my_long_string)):
# lcd_text = padded_string[((len(my_long_string)-1)-i):-i]
# mylcd.lcd_display_string(lcd_text,1)
# time.sleep(0.4)
# mylcd.lcd_display_string(padding[(15+i):i], 1)
# 
time.sleep(3)
mylcd.lcd_clear()
# =============================================================================
# Macro for special characters
# =============================================================================
#write pixel values for 5x8 grid
fontdata1 = [      
        [ 0b00010, 
          0b00100, 
          0b01000, 
          0b10000, 
          0b01000, 
          0b00100, 
          0b00010, 
          0b00000 ],
        [ 0b11111,
          0b10001,
          0b10001,
          0b10001,
          0b10001,
          0b10001,
          0b10001,
          0b11111,]
]

mylcd.lcd_load_custom_chars(fontdata1)
mylcd.lcd_write(0xC0)
mylcd.lcd_write_char(0)

mylcd.lcd_write(0x80)
mylcd.lcd_write_char(1)
mylcd.lcd_write_char(1)
mylcd.lcd_write_char(0)

# =============================================================================
# Macro for printing right pointing arrow on the left of LCD
# =============================================================================

#fontdata1 = [
#        # char(0) - Upper-left character
#        [ 0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b11111, 
#          0b11111 ],
#
#        # char(1) - Upper-middle character
#        [ 0b00000, 
#          0b00000, 
#          0b00100, 
#          0b00110, 
#          0b00111, 
#          0b00111, 
#          0b11111, 
#          0b11111 ],
#        
#        # char(2) - Upper-right character
#        [ 0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b10000, 
#          0b11000 ],
#        
#        # char(3) - Lower-left character
#        [ 0b11111, 
#          0b11111, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000 ],
#       
#        # char(4) - Lower-middle character
#        [ 0b11111, 
#          0b11111, 
#          0b00111, 
#          0b00111, 
#          0b00110, 
#          0b00100, 
#          0b00000, 
#          0b00000 ],
#        
#        # char(5) - Lower-right character
#        [ 0b11000, 
#          0b10000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000, 
#          0b00000 ],
#]
#
#mylcd.lcd_load_custom_chars(fontdata1)
#
#mylcd.lcd_write(0x80)
#mylcd.lcd_write_char(0)
#mylcd.lcd_write_char(1)
#mylcd.lcd_write_char(2)
#
#mylcd.lcd_write(0xC0)
#mylcd.lcd_write_char(3)
#mylcd.lcd_write_char(4)
#mylcd.lcd_write_char(5)